stablehlo/dialect/Base.h

```diff
@@ -73,16 +73,27 @@ std::pair<int64_t, int64_t> inferConcatenatedDimAndBound(int64_t leftSize,
                                                          int64_t leftBound,
                                                          int64_t rightBound);
 
-FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
+FailureOr<std::pair<int64_t, int64_t>> inferMostSpecificDimAndBound(
     Optional<Location> location, int64_t dim, int64_t leftSize,
     int64_t rightSize, int64_t leftBound, int64_t rightBound);
 
+FailureOr<std::pair<int64_t, int64_t>> inferLeastSpecificDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound);
+
+// Infer single least specific return type from inputTypes with support for
+// bounds. (Size, bound) of each dimension of the return type will be merged
+// from corresponding dimensions of every inputType by extracting the least
+// specific one. Return unranked tensor if any input is unranked.
+FailureOr<ShapedType> inferLeastSpecificType(Optional<Location> location,
+                                             TypeRange inputTypes);
+
 // Infer single most specific return type from inputTypes with support for
 // bounds. (Size, bound) of each dimension of the return type will be merged
-// from corresponding dimensions of every inputType by merging them.
-LogicalResult inferMostSpecificType(Optional<Location> location,
-                                    TypeRange inputTypes,
-                                    SmallVectorImpl<Type> &inferredReturnTypes);
+// from corresponding dimensions of every inputType by extracting the most
+// specific one. Return unranked tensor if all inputs are unranked.
+FailureOr<ShapedType> inferMostSpecificType(Optional<Location> location,
+                                            TypeRange inputTypes);
```

Comment:
Any particular reason to switch from LogicalResult to FailureOr? This makes the code in this PR more verbose overall, and the upside is not entirely clear (at least, not in this PR).

Reply:
We discussed a while ago to refine this function to return Type instead of passing in `SmallVectorImpl<Type> &inferredReturnTypes`, because there will only 1 inferred type returned.
Besides, this util function should handle error inside it, thus `FailureOr<ShapedType>` is the final decision. Please let me know if you have better thoughts for this.

Reply:
I don't have a strong opinion about this one. Let's proceed with what's implemented in the PR.